src/district/server/smart_contracts.cljs

```diff
@@ -115,132 +107,134 @@
-        args (if deasync? (drop-last args) args)
-        contract (instance-from-arg contract-key)
-        last-arg (last args)
-        args (if (and (map? last-arg)
-                      (not (:from last-arg)))
-               (concat (butlast args) [(merge last-arg {:from (first (web3-eth/accounts @web3))})])
-               args)
-        result (apply web3-eth/contract-call contract method args)
-        filter-id (atom nil)]
-    (if-not (fn? callback)
-      (do
-        (when (and (:print-gas-usage? @smart-contracts)
-                   (map? (last args))
-                   (string? result))
-          (handle-contract-call method result))
-        result)
-      (reset! filter-id
-              (web3-eth/filter
-               @web3
-               "latest"
-               (fn [err]
-                 (when err
-                   (callback err))
-                 (try
-                   (when (and (string? result)
-                              (map? (last args)))
-                     (loop [gas-used (handle-contract-call method result)]
-                       (when-not gas-used
-                         (recur (handle-contract-call method result (:print-gas-usage? @smart-contracts))))))
-                   (web3-eth/stop-watching! @filter-id)
-                   (callback nil result)
-                   (catch js/Error err
-                     (callback err)))))))))
-
-(def contract-call-deasynced (deasync contract-call*))
-
-(defn contract-call [& args]
-  (if (:auto-mining? @smart-contracts)
-    (apply contract-call* args)
-    (apply contract-call-deasynced args)))
+;; TODO : add alts for definite timeout (or # of retries)
+(defn- wait-for-tx-receipt*
+  "callback is a nodejs style callback i.e. (fn [error data] ...)"
+  [tx-hash callback]
+  (web3-eth/get-transaction-receipt @web3 tx-hash (fn [error receipt]
+                                                    (if error
+                                                      (callback error nil)
+                                                      (go
+                                                        (if receipt
+                                                          (callback nil receipt)
+                                                          (do
+                                                            ;; try again in 1K millis
+                                                            (<! (timeout 1000))
+                                                            (wait-for-tx-receipt* tx-hash callback))))))))
+
+(defn wait-for-tx-receipt
+  "blocks until transaction `tx-hash` gets sent to the network."
+  [tx-hash]
+  (js/Promise. (fn [resolve reject]
+                 (wait-for-tx-receipt* tx-hash (fn [error tx-receipt]
+                                                 (if error
+                                                   (reject error)
+                                                   (resolve tx-receipt)))))))
+
+(defn contract-call
```

Comment:
This is a function that we need to redesign so it supports signed transactions, if not we need to add to doc that it will only work with ganache or when using unlocked accounts

Reply:
I suggest another PR for this, since this is about removing deasync and suporting async calls?